### PR TITLE
sys-libs/libcxx: version bump and feature enhancements

### PR DESCRIPTION
**Pull request by zzlei**
This PR bumps libcxx to version 3.8.1, and also add support for libunwind and musl:
- With USE=libunwind, libcxx and libcxxrt will be built against libunwind instead of libgcc_s, thus avoiding dependence on gcc.
- When libcxx is built against musl, a simple fix will be applied.

### CODE REVIEW

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 1]
```diff
@@ -37,6 +37,11 @@ DEPEND="${RDEPEND}
 DOCS=( CREDITS.TXT )
 
 pkg_setup() {
+	if use libunwind && ! use libcxxrt ; then
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Use REQUIRED_USE for that.

File: sys-libs/libcxx/libcxx-9999.ebuild [hunk 2]
```diff
@@ -37,6 +37,11 @@ DEPEND="${RDEPEND}
...
+		eerror "Currently libunwind must be used along with libcxxrt. Building"
+		eerror "${PN} against both libsupc++ and libunwind is not supported."
```
Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: libsupc++? Wasn't this supposed to be libstdc++?

Comment (mgorny) [inline sys-libs/libcxx/libcxx-9999.ebuild]: Or wait, sorry, what wouldn't make sense ;-D.

**Comment by mgorny** [comment]
Also, doesn't libcxx support cmake? Maybe it'd take less hackery than the Makefile, considering that you're like redefining half of the build process.

**Comment by zzlei** [comment]
> Also, doesn't libcxx support cmake? Maybe it'd take less hackery than the Makefile, considering that you're like redefining half of the build process.

Yes, it does. The problem is cmake needs to know the path to llvm's root src to correctly configure, although libcxx doesn't really need any file from llvm.

The same problem also goes to other packages from llvm, especially clang. Currently clang is actually built in llvm's ebuild, because clang is so tightly coupled with llvm that it's very difficult to do a standalone build.

**Comment by mgorny** [comment]
I'm pretty sure libcxx doesn't need LLVM sources, just complains loudly about lack of them. And this can easily be worked around by `-DLLVM_CONFIG=OFF`, if I recall correctly ;-). Not saying you should do that now but if you're interested, I'd appreciate if you could try porting the ebuild to use cmake.

**Comment by zzlei** [comment]
> And this can easily be worked around by -DLLVM_CONFIG=OFF, if I recall correctly ;-)

I just did a quick check. These two settings would do (despite a warning):
`-DLLVM_CONFIG=OFF -DLIBCXX_BUILT_STANDALONE=ON`

As version 3.9 is coming out soon, I think I could rewrite the ebuild to use cmake for the new version.

**Comment by gktrk** [comment]
@aballier

**Comment by zzlei** [comment]
I'm closing this PR to combine it with my other related PRs into a new one, for the sake of easier reviewing.

**Pull request closed without merging**